### ISSUE TITLE
[Request] Colored light and variable for light falloff

### ISSUE THREAD

**Issue by DrConsole** [minetest/minetest#2670]
At the moment i'm trying to include torches and lights with other colors and area they can lit, but as I read in the forum it isn't supported by the engine yet.
So would be cool, if that wuld be included, because I think there will be more possebilities to use it than only lights.

**Comment by paramat** [comment]
Has been discussed a lot, can only happen if there is a big overhaul of the lighting system in the future.
Minetest is already fairly dark with a fast light falloff, light range cannot easily be increased and no-one wants it decreased. Increased light range would also need an overhaul of the system.

**Comment by DrConsole** [comment]
Well then I would say that time has come...
The game is pretty cool but am totaly annoyed by the dark environment...
every three blocks i have to use a torch and even right in front of the
torch its relativly dark...

**Comment by paramat** [comment]
Yes it is dark, currently light falloff must reach zero before 16 nodes (one mapblock) distance, and also falls exponentially. Also, torches were made darker a year or so ago, i think they are just a little too dark now, if you open an issue for this in MTGame you will get support.
In mods whenever i add a lighting node i make it maximum brightness (14) to compensate.

**Comment by PilzAdam** [comment]
@paramat torches are light_max-1 since 2012 (https://github.com/minetest/minetest_game/commit/f9ab8cd01838931a3f87898e6ad09c640556a1c4) and it was never changed since then.

**Comment by DrConsole** [comment]
Yeah, I also saw it in the default files, but I also tried Nodes with 14
+ light level and it wasn't satisfying as well

**Comment by paramat** [comment]
Thanks PilzAdam, weird how i thought that =/

**Comment by Zeno-** [comment]
You can change gamma in minetest.conf, which might help: https://github.com/minetest/minetest/blob/master/minetest.conf.example#L118

**Comment by paramat** [comment]
I opened an issue about meselamps not being maximum brightness https://github.com/minetest/minetest_game/issues/503 please support.

**Comment by 4aiman** [comment]
The gamma setting did nothing for me when in Win8+OpenGL...
In Linux it works perfectly.

Is it possible to make the WHOLE game a bit brighter by default w/o any shaders?

**Comment by 4aiman** [comment]
Recent build revealed that gamma effect on Linux ceased to meet my expectations. 
It's still too dark even with gamma = 1.0.
(Debian8 + XFCE)

**Comment by ProgrammingChicken** [comment]
+1, and indeed the time for a lighting overhaul is upon us.
Here's my view of the coding process btw:
More features = More people
More people = More code
More code = More bugs
More bugs  = More fixes
More fixes = More inspiration
More inspiration = More features

**Comment by C1ffisme** [comment]
I can see logic in that @ProgrammingChicken .

We've met before on Scratch, haven't we?

**Comment by ProgrammingChicken** [comment]
Why yes we have :D

**Comment by Wuzzy2** [comment]
How realistic is this feature? Any ideas on whether this could be done in an efficient manner?

**Comment by numberZero** [comment]
@Wuzzy2 Colored light can be done with just x1.5 run-time overhead, if we separate sunlight and artificial light (#5155).
Also the lighting system could be optimized, currently it seems to use (recursive) DFS which isn’t too efficient by itself, and to check many, many nodes several times. Moreover, it uses `getLight` and not `getLightRaw` or `getLightUnchecked`, although it gets the ContentFeatures then anyway...

**Comment by paramat** [comment]
Both requests are non-trivial.
Coloured light requires a new lighting system and a new map format, so very unlikely. it's been discussed at length before and no-one has yet considered it worth attempting.

Light falloff might be possible but also might be difficult since the lighting system is built around a 1 level per node falloff.

**Comment by numberZero** [comment]
Some colorization may be done at the client side. Imperfectly (due to mapblock boundaries), but that’s possible.

**Comment by raymoo** [comment]
An "easy" way to do it would be to replace scalar light values with a vector of RGB values, but the colors might look weird if you just subtract one from each component each node. For example, something like (10, 14, 10) (light green) would be come (0, 4, 0) (faint pure green) 10 nodes away. It might look ok though, I don't know.

**Comment by raymoo** [comment]
Another option is to use light colors for rendering only and not have it be part of the map lighting. Something like a point source for every lit node, though I don't know how well irrlicht handles lots of light sources.

**Comment by numberZero** [comment]
For every light source, not for every node; that may work. IrrLicht
was specially designed to support many dynamic lights. Maybe not *that*
many, though; it may be better to pre-calculate colorization in or just
before the meshgen.

**Comment by Calinou** [comment]
Even 20 dynamic lights (20 torches present in the player's view, for example, which can happen sometimes) will make rendering quite slow already.

**Comment by paramat** [comment]
In the discussions of using hardware lighting it was clear that a point source for every light source was not possible, there can be thousands of light sources in a scene (lava).so it was accepted it would be sun and moon only.

**Comment by raymoo** [comment]
Found a blog post that discusses exactly this: https://www.seedofandromeda.com/blogs/30-fast-flood-fill-lighting-in-a-blocky-voxel-game-pt-2

The artifacts from the imperfect falloff are not really that bad, and can be avoided in some ways. You can also have fancy features like light filters.

It certainly wouldn't require a full rework, just replace scalar light values with a vector of light values, and replace subtracting one from the light value with subtracting (1, 1, 1) from it. The rest of the logic is basically the same.

**Comment by raymoo** [comment]
Of course it wouldn't fit in an 8-bit param1 though, you would need to make it bigger.

**Comment by paramat** [comment]
Ok, i guess the controversy is more due to map format change and a larger 'mapnode'.
In the detailed discussions a few years ago i seem to remember that the result would be 'mapnode' doubling in size because for some reason going above 32 bits means that 64 is necessary.

**Comment by raymoo** [comment]
I don't understand why it would need to double as opposed to just have 40 bits. But from the way you phrased it I guess you don't know / disagree too.

**Comment by paramat** [comment]
Yeah i don't understand either, but the duscussion was between our most talented and knowledgable devs, so i somewhat trust them. You could search through IRC logs for dev channel.

**Comment by raymoo** [comment]
I think it may require 64 bits in memory because of alignment (unless you go through the suffering of manually working with unaligned data) though. In that case it would double the cost of map copies.

**Comment by numberZero** [comment]
On ARMs (Android), unaligned access is very problematic. On x86 it’s
easy, but slow.

OTOH, a mapblock is just 16 KiB. 16. Not 1000, not even 100. Copying
that takes 7 μs on my old PC (up to 30 μs on fresh allocation, but
still, that’s 500 MiB/s). I can’t believe that can ever be a bottleneck:
you can copy dozens of thousands of mapblocks each second (unless you
copy the metadata as well, of course).

Were there any reported cases, they could be caused either by some
synchronization problem, by the metadata, allocator slowness, etc.

**Comment by paramat** [comment]
> I can’t believe that can ever be a bottleneck

Maybe more a case of an increase in RAM use.

**Comment by raymoo** [comment]
Going by mapblock = 16 KiB though, even 1000(+ 24) mapblocks would only be 16 MiB (32 MiB if the size was doubled). Servers typically have an active mapblock range of 2 or 3, so will probably have at most 27 mapblocks active per player, and 20 players would still have much less than 1000 active blocks.

**Comment by gejan** [comment]
What about using param2?

One could use a polar approach, just store the color of the light in e.g 4 bit and mix this values proportional to the brightness:
15 values for colors = 4 shades between every primary color + White.

If 4 bits are enough und param2 is used, we can support "facedir" and "wallmounted" combined with colored light.

**Comment by numberZero** [comment]
@gejan How to mix these values when there are several colored lights around?

**Comment by ThomasMonroe314** [comment]
@gejan why change the lighting if it is just going to be re-done in the future anyways when people have time to work on it?

**Comment by gejan** [comment]
@numberZero 
Someway like this:
Assign normalized vectors to the colors e.g. red = (1, 0, 0) and the shade inbetween blue and green is 1/sqrt(2) * (0, 1, 1). Multiple by the brightness and add all the vectors from adjacent blocks. Cancell the smallest one of the rgb values and substract it from the others and normalize the result.

@ThomasMonroe314  Is there somebody who said he will work on it?

**Comment by numberZero** [comment]
@gejan I do understand how to mix two or more well-defined values encoded in your system. The question is, how to use such encoding in real lighting calculations? You will probably have to spread light from each source separately to get correct or at least consistent result.

**Comment by paramat** [comment]
Any dev support for this feature?

**Comment by Ezhh** [comment]
The issue might be "old" but it's still a very much wanted thing...

**Comment by paramat** [comment]
Sorry, label wrongly added.

**Comment by numberZero** [comment]
![_2018-03-23_02-02-03](https://user-images.githubusercontent.com/7352626/38517363-6eb41988-3c42-11e8-8121-fbb98030f213.png)

**Comment by rubenwardy** [comment]
Nice! What performance cost will making it into 3D have?

**Comment by numberZero** [comment]
@rubenwardy It’s 3D) It shows a plane section of a 3D voxel world)
The performance (per light source) is good, after program initialization (which takes some time, but that data could be precomputed), but I use lighting model radically different from MT, that is unsuitable for MT. Still, the problem in MT is not performance itself, but the hassle of changing the whole system IMO.

**Comment by rdococ** [comment]
As a compromise, you could have two light values. One for warm sources of light, and one for cool sources of light. Think of them as YB, yellow and blue, instead of the full RGB, but you could color them any colors you want.

**Comment by benrob0329** [comment]
Couldn't this be implemented as a shader option?

**Comment by ClobberXD** [comment]
While coloured lighting involves fragment shader modifications, it can't be implemented solely on the shader-side - most importantly, we actually need the 2x more light values per-node to be stored somewhere (preferably in the map itself). AFAIK, the main/only hurdle in implementing this is the compat breakage in map/world-format.

**Comment by TumeniNodes** [comment]
> Couldn't this be implemented as a shader option?

again though @benrob0329 we have enough shaders which need to be tweaked as it is. It makes no sense to add new shaders until the existing ones are sorted out.

https://learnopengl.com/Lighting/Basic-Lighting
there is _a lot_ of work involved

>So I'm for it as long as it's done properly.

But I recall RBA mentioning in order to fix these issues... (the right way) would mean a ton of tearing things apart and causing serious breakage.
The longer it is put off, the worse/more work it will take but, at the same time... "serious breakage"

If we look at all of this rationally... taking into consideration the skeleton team and even those present's lack of time to do a whole lot... this sort of issue seems more and more unrealistic.

**Comment by ClobberXD** [comment]
> there is a lot of work involved

Yes, a lot of work is involved in implementing lighting from scratch. But MT has already implemented single-colour lighting, and most of the code is already in place. Implementing coloured lighting is just about adding support for it in the Lua API, and tweaking some code (incl. shaders) to work with three separate channels instead of one value for all channels.

**Comment by benrob0329** [comment]
My point was more that if the lighting was drawn by the shader code, then the color values for nodes could be passed to it and applied as appropriate without an update to the map format. It'd be strictly visual, but that's the main point of this FR. no?

**Comment by ClobberXD** [comment]
> the color values for light emitting nodes could be passed to it and applied as appropriate without an update to the map format

Where would the values be passed from? Where would they be stored? :)

**Comment by benrob0329** [comment]
>Where would the values be passed from?

The CPP code, in the same way any other value might be passed to the GPU. I could attempt to guess at how it could be implemented, but without being familiar with MT's rendering code I'd probably be way off.

>Where would they be stored?

Well if the light colors are stored in the nodedef (as these won't change) then you already have a method of getting them, you'll just need to make it quick to access. An array indexed by the content ID of each node with the appropriate lighting color code might work, but this is a stab in the very dim, uncolored light.

The current shader code (and the fixed pipeline iirc) already colors the light based on whether it's moonlight, sunlight, or light emitted by a node. I can't imagine it'd be that much of a stretch to alter this logic to color the light coming from a node of type X in a similar way. But again, this is a shot in the dark.

I really don't feel that the color of the light needs to be stored in the map, just applied in the rendering code based on the type of node it's being emitted from.

**Comment by paramat** [comment]
> most of the code is already in place. Implementing coloured lighting is just about adding support for it in the Lua API, and tweaking some code (incl. shaders) to work with three separate channels instead of one value for all channels.

No, it's not as simple as you make it sound. It's a huge change to MT that would require increasing the size of the map node to contain the extra light information, so map format would break, old worlds would need conversion code run on them to slowly convert them during use.
Memory use in the engine and in world databases would increase significantly.
Performance would drop significantly because the lighting code is already very intensive.
Plus of course many unforseen problems and bugs caused.

If it happens it would be MT 6.0.0 and many years away, and only if we somehow gain much more core dev time. However even with much more core dev time there are many far higher priorities.

Light colour does need to be stored in the map, just like light is currently, that's how MT works.

**Comment by benrob0329** [comment]
Unless the color of the light is going to be important for mods to know (which still could be done via checking the nodedef) I really doubt it _must_ be stored in the map, and not just applied via some logic and a reference array with a shader. If it must it must, but I don't feel like anyone knowledgeable has looked into the possibility of it not needing to.

**Comment by TumeniNodes** [comment]
all the magic is done within the engine... but the engine must get information from the map. So both are key components of the overall system.
There is no way around that. It is the same for every game/engine which uses such visual graphics

**Comment by benrob0329** [comment]
>There are "shader mods" but all they really do, is look at the engine for information... which the engine then still has to get from the map

You're missing the point, and acting like I know nothing about how computer graphics work. Shaders take information passed to the GPU, and an array with information on how to color the lighting based on what node is emitting it does not need to be a part of the map. If you take the map information (what node is it?) and combine that with some info from NodeDef (what color should it's light be?) you can (theoretically) draw the lighting appropriately. You don't need to have the exact color stored in the map (unless someone actually says why, other than "it can't").

>All these processes are fine for those who have equipment beefy enough to handle it... but sadly, most do not... Most end users do not have top of the line/state of the art devices... which makes such visuals useless to them
It is only the top-end gaming community who benefits... (people with $1500+ rigs) some of them that is the cost of just their graphics card and GPU 0_o

You are completely misrepresenting what shader code is based on the limited (and backwards) way Minetest works. Shaders are GPU programming, nothing more. They can be as lightweight, or as heavy as you make them. **GPUs need code to display graphics.** Shaders are not inherently more complex or difficult to run, they are required (even if the GPU runs it's own code if the CPU doesn't provide it with any of it's own) on modern hardware.

**Comment by TumeniNodes** [comment]
> acting like I know nothing about how computer graphics work.

No at all, I apologize if I'm coming off this way.
I promise, you know more and understand more about it than I do, I was merely going off what I know related to how shaders work... and my bit of knowledge is most likely quite outdated.

To be clear and honest, _if_ someone is able to provide much better shaders here, I am all for it.
I'd still love to see them... I just don't have my hopes up for it any time soon for a number of reasons.
If it has to be performed in a patchy/hacky manner in any way, then I am against it.

As far as I have seen in the past, when these bits were added in/coded, they were not done correctly for the most part, and as time has moved on, those issues become buried deeper and harder to dig out and correct without impacting other various areas, creating a pandora's box.

If people have not done so yet, I suggest it may be a good idea to take some time to go back in logs and PRs/Issues related to RBA. It all still provides good information.
He argued and stressed over the very backwardness you mention, and pushing to fix it. The roadblock was breakage.. in many areas.
Which is understandable given the limited manpower existing right now, even with contributors.

If someone gets into this, and along the way finds it is more than they thought... it can bring some serious problems.
This has to be something those proposing to work on it are absolutely sure and certain and are looking ahead

I did have someone I know have a look at them once who stated "it's a mess".
Sadly that individual has not the time nor the interest to get involved.

My comments related to this are mostly negative, and everyone here has a far better understanding than myself but, I know some basic things and realized early on when I first arrived that the shaders situation is sketchy at best.
Id love to see basic shadows working with the sun & moon, and I'd love to have  a lighting system which at the very least does not bleed through walls, etc..
Those are really the only two things I truly want from a visuals/graphics standpoint.

**Comment by ClobberXD** [comment]
@benrob0329 Ah now I see where you're going with this. Your suggestion would prevent compat breakage of map/world format, which is great. But then, mods should know the colour of the light as well... or, do they? :thinking:

****

Wall of text incoming; apologies in advance :)

> It's a huge change to MT that would require increasing the size of the map node to contain the extra light information, so map format would break, old worlds would need conversion code run on them to slowly convert them during use.

That's true, and I did mention that in one of my previous comments:

> AFAIK, the main/only hurdle in implementing this is the compat breakage in map/world-format.

> Memory use in the engine and in world databases would increase significantly.

I agree that coloured lighting increases memory consumption, but not by much actually. Here's a (rough approximation) breakup of the contents of a map block:

- Map format version number (1B)
- Lighting flags (1B)
- `lighting_complete` (2B) 
  - This seems to be a bunch of flags to determine if the mapblock edges are correctly lit, and the lighting calculations are performed otherwise.
- `content_width` (1B)
  - This stores the width of `param0` (i.e. the node's content ID AFAIK)
- Size of an individual node (1B)
- zlib-compressed node data (16 * 16 * 4 or 1024B, **when uncompressed**)
  - Each node is 4B, and `16 * 16 * 4` is the uncompressed size in bytes.
  - param0 (2B)
  - param1 (1B)
  - param2 (1B)
- Node meta (variable size; can technically even be in the order of megabytes)

Each map-block is exactly 1030B (1024B node data + 6B other data) in size. Note that this figure ignores node meta - adding node meta into the mix can make each block twice or thrice as large on average (just making a guess here). Also note that this calculation, for a brief second, completely dismisses the fact that nodes are compressed before storing.

Currently, `param1` (1B / 8b) stores lighting like so:
- Day light (4b)
- Night light (4b)

Now, to replace one channel with three, we need to add 16 more bits per-node (2 lights stored (day and night) * 2 extra channels * 4 bits per-channel). Each node would then become 6B (4B + 2B) in size. So now, the total size of nodes in the mapblock will be **(16 * 16 * 6) or 1536B**. If the mapblock lighting flags needn't be altered, the size of the mapblock would now be **1542B** (1536B node data + 6B misc. data).

While that is a significant increase in required memory, that's actually not much when taking into consideration the facts that the calculation ignores node data compression, and doesn't give a damn node meta - what's a couple of bytes when the mapblock's size is already in the order of kilobytes (or even megabytes)? :D

> Performance would drop significantly because the lighting code is already very intensive.

While I can't say for sure, the rendering side of things (shaders, for example) should already consider light as a 3-component colour. We'd only have to allow mods to customise all three channels of the colour, instead of scaling the colour on a greyscale based on `light`. The performance drop certainly wouldn't be significant (I hope). :)

**Comment by Ekdohibs** [comment]
Actually, a total of 2B per node of lighting would be enough: 4b for natural light (sunlight/moonlight), and 4b*3 for artificial light. Then, actual light can be extracted from that and the knowledge of the color of sunlight or moonlight, which could even be made configurable.
However, there is a slight error in your calculations: there are 16^3 nodes in a mapblock, that is, 4096, so the uncompressed node data already uses 16kB memory (and with the optimization I'm proposing, 4kB additional memory would be needed).

**Comment by ClobberXD** [comment]
> However, there is a slight error in your calculations: there are 16^3 nodes in a mapblock, that is, 4096, so the uncompressed node data already uses 16kB memory

Thanks for pointing that out, I'll fix the calculations right away. :)

> Actually, a total of 2B per node of lighting would be enough: 4b for natural light (sunlight/moonlight), and 4b*3 for artificial light. Then, actual light can be extracted from that and the knowledge of the color of sunlight or moonlight, which could even be made configurable.

Oh nice! This seems to make more sense than the goofy calculations I came up with.

**Comment by DS-Minetest** [comment]
>Currently, `param1` (1B / 8b) stores lighting like so:
>- Day light (4b)
>- Night light (4b)

I don't know much about how `param1` stores light values. I only know that it's a little bit weird.
But day light and night light is the same, so this here makes not much sense, could you explain this in more detail?

>But would sunlight/moonlight be colourable then?

Why would it be needed to store the color of sunlight or moonlight per node? There's just a big lamp directly above the minetest world that produces parallel light shining down. And the color depends on the time.

What @benrob0329 suggested sounded like if the client should search for the light emitting nodes and show the color according to the value in the emitting nodes' nodedef.

If param1 was changed to use 2 bytes, it might make sense to give param2 2B, too, and interpret the node id as _unsigned_ short.

The size of the world is not the only problem with multiple colors in the light, another problem would be that the lightning spread calculations would have to be done 3 times. (I don't know if it's really that bad, maybe the "calculations" are much faster than the mapsaving for the updated nodes.)

**Comment by ClobberXD** [comment]
> I don't know much about how param1 stores light values. I only know that it's a little bit weird.
But day light and night light is the same, so this here makes not much sense, could you explain this in more detail?

I'm sorry, I have zero idea about our lighting. :)

> Why would it be needed to store the color of sunlight or moonlight per node? There's just a big lamp directly above the minetest world that produces parallel light shining down. And the color depends on the time.

Oh right, good point. So the 4b for natural light would only hold the intensity.

> another problem would be that the lightning spread calculations would have to be done 3 times

That's not really the case. MT currently hard-codes the light to be of a white colour, but that's it. The shader still treats it as a `vec3`. "Un-hardcoding" the colour of the light would have zero performance impact in this regard. See:

https://github.com/minetest/minetest/blob/ded5da780002c17f2079a1d8ea09eb923a3b5e8f/client/shaders/nodes_shader/opengl_vertex.glsl#L18-L19

This colour is then used to calculate the final colour of the fragment here:

https://github.com/minetest/minetest/blob/ded5da780002c17f2079a1d8ea09eb923a3b5e8f/client/shaders/nodes_shader/opengl_vertex.glsl#L125-L142

**Comment by DS-Minetest** [comment]
> > another problem would be that the lightning spread calculations would have to be done 3 times
> 
> That's not really the case. MT currently hard-codes the light to be of a white colour, but that's it. The shader still treats it as a `vec3`. "Un-hardcoding" the colour of the light would have zero performance impact in this regard. See:
…

I wasn't referring to anything shader related.
If you place a eg. torch, for every node near that has `paramtype1="light"` and less light, the prarm1 has to be updated, this happens recursively.

**Comment by ClobberXD** [comment]
> If you place a eg. torch, for every node near that has paramtype1="light" and less light, the param1 has to be updated, this happens recursively.

I haven't been to that part of the code yet, but I don't see how coloured lighting would increase the number of calculations performed on each light update. Also, those calculations only determine the brightness/light intensity of the affected nodes, and not the colours, AFAIK.

**Comment by DS-Minetest** [comment]
> I haven't been to that part of the code yet

Me neither (just to clarify, hehe).
> but I don't see how coloured lighting would increase the number of calculations performed on each light update. Also, those calculations only determine the brightness/light intensity of the affected nodes, and not the colours, AFAIK.

The brightness of each color (green, blue and red) would have to spread.

**Comment by ClobberXD** [comment]
> The brightness of each color (green, blue and red) would have to spread.

I'm not sure that's how it works. I think it's more like the intensity is calculated based on distance from light source(s), and the final colour is multiplied with the (normalised) intensity in the shader.

**Comment by Ekdohibs** [comment]
After some more thinking, I noticed that light as it is currently done wouldn't allow things like colored glass, that let light pass through but giving it some tint. The lighting system would probably need to be rethought completely if we want to handle that.

**Comment by rdococ** [comment]
The brightness of every translucent or transparent node is calculated by spreading recursively from the light sources. The color of the light at each position would also need to be calculated this way to handle color mixing properly.

One of the more important problems I think is that a better algorithm for spreading light needs to be found. A [breadth-first algorithm](https://www.seedofandromeda.com/blogs/29-fast-flood-fill-lighting-in-a-blocky-voxel-game-pt-1) would work, but I'm not sure if Minetest already uses it or not.

**Comment by numberZero** [comment]
> would have to be done 3 times

4 times (sunlight and R/G/B artificial), instead of current 2 (day and night). Or 6 times (R/G/B both for sunlight and artificial).

> The color of the light at each position would also need to be calculated this way to handle color mixing properly.

That’s done trivially by treating each channel independently. The solution may not be visually perfect, although it is correct assuming the light level scale is logarithmic (it is documented to be, somewhere). This way, color tint may be added trivially as well, if you use RGB both for artificial and natural light (or for day and night light).

One interesting problem is compatibility. There was an attempt to separate natural and artificial light (#5687), it wasn’t completed. Without that, only change to 2*RGB (day+night) model would be non-breaking. On the other hand, new system may be introduced for new worlds only (offline conversion will always be possible, as long as node definitions are known; there are no such tools currently, though).

> One of the more important problems I think is that a better algorithm for spreading light needs to be found.

For speed, many improvements are possible. For example:
1. Stop using nibbles in favor of full bytes. CPU can’t work with nibbles (except of the dreaded BCD arithmetic). That would double memory usage for lighting, though.
2. Store more data in the node, to save accesses to the nodedef. Just a few bits (like light_propagates, is_light_source) would save tons of such calls, and could fit even in content_id (have you ever seen 65k nodes?). Or replace padding in the grown node structure.
3. Use LTO (link-time optimization) to inline `MapNode::setLight` =)

But remember that for any optimization, performance gains must be measured, they may not be as expected.

> A breadth-first algorithm would work, but I'm not sure if Minetest already uses it or not.

It does, but in the presence of cache, it may not be the best choice.

**Comment by SmallJoker** [comment]
@paulevsGitch And how do you store the "night" light information? Sunlight changes depending on game time, thus you'd probably need two more bytes to interpolate the light for light-propagating non-light-source nodes. Either that or spend CPU time recalculating that with each change.

**Comment by rubenwardy** [comment]
You don't need to store night light, as it is exactly equal to artificial light

**Comment by paulevsGitch** [comment]
If only 12 bits will be used the sunlight can be stored in upper 4 bits, it’s color and intensity will depend on daytime

**Comment by paulevsGitch** [comment]
This can also be used for smooth day/night cycle if sunlight will not be baked into geometry and will be calculated from time and intensity (stored in upper 4 bits).

**Comment by ryvnf** [comment]
There would still be a point to store natural light as RGB channels. Another application of separating the light channels is that you can have colored transparent nodes which block certain channels. Like making white light passed through a red window red light. It would be a bit inconsistent if one could make a building with a red glass roof, but the natural light show up as white, even though it should be red due to the glass roof. Would probably also cause some ugly light blending if there is an artificial light source on the roof as well.

**Comment by ryvnf** [comment]
I do not think this is worth adding at the moment because it requires a lot of changes to the code and the map format. Adding RGB lighting would probably also mean that each light channel has to be distributed separately for white lighting. In practice that would be a 3x slowdown for all white light distribution. Since white light most likely would remain the most common form of light, that is an expensive price to pay for the few circumstances where colored light would be useful.

I personally think a more realistic and interesting approach would be to expose a way to change the color palettes of natural and artificial light in Lua. But that would be a different feature request I guess.

**Comment by ezileicorreia** [comment]
A lot of time is wasted trying to solve problems that don't yet exist. Color light doesn't have to be something extraordinary, a function that turns white light into RGB light would solve the problem. We just need a testable code.

**Comment by ezileicorreia** [comment]
I found a lighting tutorial, hope it's useful and someone is willing to put it into practice. We'll never know if that's possible until we have testable color lighting. I could make a list of things I'd like to see in Minetest, but lighting is just as fundamental as mapping shadows. Hope this helps.

https://learnopengl.com/Lighting/Colors

**Comment by ezileicorreia** [comment]
Maybe it helps! https://nnarain.github.io/2015/07/23/Voxel-Lighting!.html

**Comment by DrConsole** [comment]
Merging my accounts
In future I will be available as "Jan-NiklasB"

**Comment by Jan-NiklasB** [comment]
Here we go again

**Comment by Froggo8311** [comment]
Now that we have the new dynamic lighting engine, perhaps this might be easier to implement now

**Comment by rubenwardy** [comment]
No, dynamic shaders are purely shader based and only works with the sun/moon. It hasn't made colored lighting easier to implement, as that will need the same changes as before to the voxel lighting engine

**Comment by numberZero** [comment]
> Would it be easier to add coloured lighting to the game via the help of shaders?

It would be (for smooth day/night transition) but apparently Minetest doesn’t support shaders. OpenGL 2-level things don’t count.

And of course #5155 would be helpful albeit not mandatory.

**Comment by farooqkz** [comment]
I think this needs to be closed as the author is not there anymore, and we can't use the code cuz licencing problems?

**Comment by appgurueu** [comment]
I see no need to close this: It's a popular, concept-approved feature request, not a PR.

It's not really relevant whether there exists any usable code for it yet. Did you maybe intend to comment on some PR instead?

**Comment by farooqkz** [comment]
> I see no need to close this: It's a popular, concept-approved feature request, not a PR.
> 
> It's not really relevant whether there exists any usable code for it yet. Did you maybe intend to comment on some PR instead?

Sorry I thought it's a PR :)
